### PR TITLE
feat(gui-client): make all modules Linux-friendly

### PR DESCRIPTION
**Pull request by ReactorScram**
Splits up platform-specific modules into `linux.rs` and `windows.rs` submodules, or `imp` modules within the parent module, so that we can compile all the platform-independent code (e.g. Tauri calls, `run_controller`) for Linux.

This will show the Tauri GUI on Linux:

image

### CODE REVIEW

File: rust/gui-client/src-tauri/Cargo.toml [hunk 1]
```diff
@@ -18,47 +18,48 @@ clap = { version = "4.4", features = ["derive",  "env"] }
...
-rand = "0.8.5"
-windows-implement = "0.52.0"
-sadness-generator = "0.5.0"
-bincode = "1.3.3"
-native-dialog = "0.7.0"
-output_vt100 = "0.1"
 
-# These dependencies are locked behind `cfg(windows)` because they either can't compile at all on Linux, or they need native dependencies like glib that are difficult to get. Try not to add more here.
+[target.'cfg(linux)'.dependencies]
+dirs = "5.0.1"
```
Comment (ReactorScram) [inline rust/gui-client/src-tauri/Cargo.toml:56]: This should have already been in #3734, I'm not sure how it kept getting missed.

File: rust/gui-client/src-tauri/Cargo.toml [hunk 2]
```diff
@@ -18,47 +18,48 @@ clap = { version = "4.4", features = ["derive",  "env"] }
...
 # Same crate Hickory uses
 hostname = "0.3.1"
 # This is the same crate hickory uses to get system resolvers
 ipconfig = "0.3.2"
 keyring = "2.3.2"
 known-folders = "1.1.0"
 minidumper = "0.8.1"
+native-dialog = "0.7.0"
+output_vt100 = "0.1"
+rand = "0.8.5"
```
Comment (ReactorScram) [inline rust/gui-client/src-tauri/Cargo.toml]: Some of these aren't new, I just alphabetized them.

File: rust/gui-client/src-tauri/src/client/debug_commands.rs [hunk 3]
```diff
@@ -48,15 +46,3 @@ fn hostname() -> Result<()> {
...
-fn wintun() -> Result<()> {
-    tracing_subscriber::fmt::init();
-    let path = connlib_shared::windows::wintun_dll_path()?;
-    // SAFETY: Loading a DLL from disk is unsafe. We're responsible for making sure
-    // we're using the DLL's API correctly.
-    unsafe { wintun::load_from_path(&path) }?;
-    tracing::info!(?path, "Loaded wintun.dll");
-
-    Ok(())
-}
```
Comment (ReactorScram) [inline rust/gui-client/src-tauri/src/client/debug_commands.rs:62]: I'll rewrite this later if I ever need it again.

File: rust/gui-client/src-tauri/src/client/deep_link.rs [hunk 4]
```diff
@@ -1,43 +1,56 @@
...
-use connlib_shared::{control::SecureUrl, BUNDLE_ID};
+use connlib_shared::control::SecureUrl;
 use secrecy::{ExposeSecret, Secret, SecretString};
-use std::{ffi::c_void, io, path::Path, str::FromStr};
-use tokio::{io::AsyncReadExt, io::AsyncWriteExt, net::windows::named_pipe};
-use windows::Win32::Security as WinSec;
+#[cfg(target_os = "windows")]
+use std::io;
 
-pub(crate) const FZ_SCHEME: &str = "firezone-fd0020211111";
```
Comment (ReactorScram) [inline rust/gui-client/src-tauri/src/client/deep_link.rs:11]: This should actually stay up here and be cross-platform.

File: rust/gui-client/src-tauri/src/client/deep_link.rs [hunk 5]
```diff
@@ -1,43 +1,56 @@
...
+#[cfg(target_os = "windows")]
+#[path = "deep_link/windows.rs"]
+mod imp;
 
 #[derive(thiserror::Error, Debug)]
 pub enum Error {
     #[error("named pipe server couldn't start listening, we are probably the second instance")]
     CantListen,
     /// Error from client's POV
+    #[cfg(target_os = "windows")]
```
Comment (ReactorScram) [inline rust/gui-client/src-tauri/src/client/deep_link.rs:22]: Should I try to pile all these into a `WindowsError` or something? Or just make them into `anyhow`?

File: rust/gui-client/src-tauri/Cargo.toml (hunk 1 above)
Comment (ReactorScram) [reply rust/gui-client/src-tauri/Cargo.toml:56]: It was down in a Linux-specific section.

File: rust/gui-client/src-tauri/src/client/crash_handling.rs [hunk 6]
```diff
@@ -118,7 +124,11 @@ impl minidumper::ServerHandler for Handler {
             .expect("Should be able to find logs dir to put crash dump in")
             .join("last_crash.dmp");
 
+        // `tracing` is unlikely to work inside the crash handler subprocess, so
+        // just print to stderr and it may show up on the terminal. This helps in CI / local dev.
+        eprintln!("Creating minidump at {}", dump_path.display());
         if let Some(dir) = dump_path.parent() {
+            // TODO: There is a TOCTOU bug here
             if !dir.try_exists()? {
```
Comment (conectado) [inline rust/gui-client/src-tauri/src/client/crash_handling.rs]: do we need to check if this exists? I think `create_dir_all` does nothing if it exists

File: rust/gui-client/src-tauri/src/client/deep_link.rs [hunk 7]
```diff
@@ -1,43 +1,60 @@
...
+mod imp;
+
+// TODO: The repeated `target_os = "windows"` here is bad.
+// When I add deep link support for Linux I'll replace this all for `anyhow`.
 #[derive(thiserror::Error, Debug)]
 pub enum Error {
     #[error("named pipe server couldn't start listening, we are probably the second instance")]
     CantListen,
     /// Error from client's POV
+    #[cfg(target_os = "windows")]
```
Comment (conectado) [inline rust/gui-client/src-tauri/src/client/deep_link.rs]: until we replace this with anyhow can't we just have this for both platforms, we can allow dead code for linux for this enum if the compiler complains

File: rust/gui-client/src-tauri/src/client/deep_link/linux.rs [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+use connlib_shared::control::SecureUrl;
+use secrecy::Secret;
+
+pub(crate) struct Server {}
+
+impl Server {
+    pub(crate) fn new() -> Result<Self, Error> {
+        tracing::warn!("Not implemented yet");
+        tracing::trace!(scheme = super::FZ_SCHEME, "prevents dead code warning");
+        // Stop Cargo from erroring
```
Comment (conectado) [inline rust/gui-client/src-tauri/src/client/deep_link/linux.rs]: what's the error here?

File: rust/gui-client/src-tauri/src/client/gui.rs [hunk 9]
```diff
@@ -55,6 +63,7 @@ impl Managed {
 // TODO: Replace with `anyhow` gradually per <https://github.com/firezone/firezone/pull/3546#discussion_r1477114789>
 #[derive(Debug, thiserror::Error)]
 pub(crate) enum Error {
+    #[cfg(target_os = "windows")]
```
Comment (conectado) [inline rust/gui-client/src-tauri/src/client/gui.rs]: same as above about errors

File: rust/gui-client/src-tauri/src/client/network_changes.rs [hunk 10]
```diff
@@ -1,355 +1,9 @@
...
-        Networking::NetworkListManager::{
-            INetworkEvents, INetworkEvents_Impl, INetworkListManager, NetworkListManager,
-            NLM_CONNECTIVITY, NLM_NETWORK_PROPERTY_CHANGE,
-        },
-        System::Com,
-    },
-};
+#[cfg(target_os = "windows")]
+#[path = "network_changes/windows.rs"]
+mod imp;
```
Comment (conectado) [inline rust/gui-client/src-tauri/src/client/network_changes.rs:7]: what does `imp` stand for?

**Review by conectado — COMMENTED** [review]
Looks quite good! I just have some clarifying questions

File: rust/gui-client/src-tauri/src/client/network_changes.rs (hunk 10 above)
Comment (ReactorScram) [reply rust/gui-client/src-tauri/src/client/network_changes.rs:7]: "Implementation". I'm open to better names, I just keep waffling between "impl" "detail" and "inner" for these things where I can't think of anything descriptive

File: rust/gui-client/src-tauri/src/client/deep_link/linux.rs (hunk 8 above)
Comment (ReactorScram) [reply rust/gui-client/src-tauri/src/client/deep_link/linux.rs]: Dead code. It should go away once I allow dead code for that Error enum

**Review by conectado — APPROVED** [review]
LGTM

**Pull request merged**